src/Middleware/RetryHandler.cs

```diff
@@ -248,5 +230,37 @@ private static bool ShouldRetry(HttpStatusCode statusCode)
                 _ => false
             };
         }
+
+        private static async Task<Exception> GetInnerException(HttpResponseMessage response, CancellationToken cancellationToken)
+        {
+            var httpStatusCode = response.StatusCode;
+            string? errorMessage = null;
+
+            // Drain response content to free connections. Need to perform this
+            // before retry attempt and before the TooManyRetries ServiceException.
+            if(response.Content != null)
+            {
+#if NET5_0_OR_GREATER
+                var responseContent = await response.Content.ReadAsByteArrayAsync(cancellationToken).ConfigureAwait(false);
+#else
+                var responseContent = await response.Content.ReadAsByteArrayAsync().ConfigureAwait(false);
+#endif
+                errorMessage = GetStringFromContent(responseContent);
+            }
+
+            return new Exception($"HTTP request failed with status code: {httpStatusCode}.{errorMessage}");

```

Comment:
```suggestion
            return new ApiException($"HTTP request failed with status code: {httpStatusCode}.{errorMessage}")
            {
                ResponseStatusCode = (int)response.StatusCode,
                ResponseHeaders = response.Headers.ToDictionary()
            };
```

This way we can leverage known properties in the exception class and add more error info as the exception class evolves as captured in https://github.com/microsoft/kiota-http-dotnet/issues/206.